src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -0,0 +1,244 @@
+    Transition_frontier.clear_paths frontier ;
+    don't_wait_for (on_transition t ~ledger_hash_table (transition, tm)) ;
+    let%map tree = Syncable_ledger.valid_tree t.syncable_ledger in
+    let root_hash = Ledger.merkle_root tree in
+    let state_hash = Hashtbl.find_exn ledger_hash_table root_hash in
+    Transition_frontier.rebuild frontier tree state_hash ;
+    Hashtbl.clear ledger_hash_table ;
+    toggle_pipes ()
+
+  let create ~parent_log ~frontier ~network ~ancestor_prover =
+    let logger = Logger.child parent_log __MODULE__ in
+    let initial_breadcrumb = Transition_frontier.root frontier in
+    let initial_root_state =
+      initial_breadcrumb |> Transition_frontier.Breadcrumb.transition_with_hash
+      |> With_hash.data |> External_transition.protocol_state
+    in
+    let ledger =
+      Transition_frontier.Breadcrumb.staged_ledger initial_breadcrumb
+      |> Staged_ledger.ledger
+    in
+    { network
+    ; logger
+    ; ancestor_prover
+    ; best_with_root= {state= initial_root_state; root= initial_root_state}
+    ; syncable_ledger= Syncable_ledger.create ledger ~parent_log:logger }
+
+  let run ~valid_transition_writer ~processed_transition_writer
+      ~catchup_job_writer ~catchup_breadcrumbs_writer ~parent_log ~network
+      ~ancestor_prover ~frontier ~transition_reader =
+    let toggle_pipes () =
+      Closed_writer.(
+        toggle valid_transition_writer ;
+        toggle processed_transition_writer ;
+        toggle catchup_job_writer ;
+        toggle catchup_breadcrumbs_writer)
+    in
+    let is_bootstrapping () =
+      let are_pipes_closed =
+        Closed_writer.(
+          is_closed valid_transition_writer
+          && is_closed processed_transition_writer
+          && is_closed catchup_job_writer
+          && is_closed catchup_breadcrumbs_writer)
+      in
+      Debug_assert.debug_assert (fun () ->
+          assert (
+            Closed_writer.(
+              are_pipes_closed
+              || not
+                   ( is_closed valid_transition_writer
+                   || is_closed processed_transition_writer
+                   || is_closed catchup_job_writer
+                   || is_closed catchup_breadcrumbs_writer )) ) ) ;
+      are_pipes_closed
+    in
+    let t = create ~parent_log ~frontier ~network ~ancestor_prover in
+    let ledger_hash_table = Ledger_hash.Table.create () in
+    Reader.iter transition_reader
+      ~f:(fun (`Transition incoming_transition, `Time_received tm) ->
+        let new_transition = Envelope.Incoming.data incoming_transition in
+        let root_transition =
+          Transition_frontier.root frontier
+          |> Transition_frontier.Breadcrumb.transition_with_hash
+          |> With_hash.data
+        in
+        let root_state = External_transition.protocol_state root_transition in
+        let new_state = External_transition.protocol_state new_transition in
+        if should_bootstrap root_state new_state then
+          don't_wait_for
```

Comment:
rather than `don't_wait_for`-ing to feed the bootstrap things, I think you should use a Strict_pipe for communication here (and therefore split out that bit into a separate module). As written, we'll be spinning up a "green-thread" on every iteration and thus can be DoS-ed by being sent too many transitions

Reply:
If you do that, I think you don't need to have this module have any `t` value nor any sort of create -- that can be entirely managed by the `run` of some other module you make to handle the new data coming in here.